DHT/dht.c

```diff
@@ -877,21 +885,26 @@ void dhtRemoveElement(HashTable *ht, dhtConstValue key)
   TraceFunctionResultEnd();
 }
 
-dhtElement *dhtEnterElement(HashTable *ht, dhtConstValue key, dhtConstValue data)
+dhtElement *dhtEnterElement(HashTable *ht, dhtKey key, dhtValue data)
 {
   InternHsElement **phe, *he;
-  dhtConstValue KeyV;
-  dhtConstValue DataV;
+  dhtKey KeyV;
+  dhtValue DataV;
 
   TraceFunctionEntry(__func__);
   TraceFunctionParam("%p",(void *)ht);
-  TraceFunctionParam("%p",(void const *)key);
-  TraceFunctionParam("%p",(void const *)data);
+#if (defined(__cplusplus) && (__cplusplus >= 201103L)) || (defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L))
+  TraceFunctionParam("%jx",key.value.unsigned_integer);
+  TraceFunctionParam("%jx",data.unsigned_integer);
+#else
+  TraceFunctionParam("%lx",(unsigned long)key.value.unsigned_integer);
+  TraceFunctionParam("%lx",(unsigned long)data.unsigned_integer);
+#endif
   TraceFunctionParamListEnd();
 
-  assert(key!=0);
-  KeyV = (ht->procs.DupKey)(key);
-  if (KeyV==0)
+  assert(key.value.object_pointer!=0); /* TODO: This assert assumes that object_pointer is the active member.
+                                          Is there a more generic test we could do?  Do we need one? */
+  if ((ht->procs.DupKeyValue)(key.value, &KeyV.value))
```

Comment:
Hmm.
Before, we signaled "key duplication failed" if DupKey returned 0.
Now, we signal it if DupKeyValue doesn't.
Can this be correct?

Reply:
It's admittedly confusing.&nbsp; Previously we `assert`ed that `key` is non-`NULL`, so now I `assert` that `key`'s stored object pointer is non-`NULL`.&nbsp; I've replaced the previous call to `DupKey` with a call to `DupKeyValue`, which is specified to return 0 on success and nonzero on failure (e.g., memory exhaustion), hence the nonzero return value corresponds to `DupKey` returning `NULL`.&nbsp; (`DupKeyValue` _is_ allowed [and supposed] to "duplicate" a `NULL` object pointer and return 0; this is needed elsewhere in this file.)